File: GVO Code v3/Slew.pas
unit Slew;

interface
uses utils, sysutils,windows,dialogs;

Procedure SlewScope;

implementation
uses Main,quad1to1,quad1to3,quad3to3,quad3to1,
     quad1to2, quad2to1, quad2to2,GetQandYU ;

Procedure SlewScope;
label endit;
var
 i:integer;
 Alt,HA,Xcount,Ycount : real;
 thandle : HWND;
 Begin
   thandle := 0;
   mainform.sjog.enabled := False;
   mainform.njog.enabled := False;
   mainform.ejog.enabled := False;
   mainform.wjog.enabled := False;
   RaTarget := Coord(pndomem^).RAGoTo ;
   DecTarget := Coord(pndomem^).DECGoTo ;
   GetQandY(RaTarget,DecTarget,Alt,HA,Xcount,Ycount,TargetQuadrant,i);
   if(Alt < 20.0) then
   begin
      Coord(pndomem^).RAGoTo := 0.0   ;
      Coord(pndomem^).decGoTo := 0.0  ;
      windows.MessageBox(thandle,'No slew below horizon','Telescope Control',MB_SYSTEMMODAL);
      goto endit;
   End;
   if(i= 0) then
   begin
      Coord(pndomem^).RAGoTo := 0.0   ;
      Coord(pndomem^).decGoTo := 0.0  ;
      windows.MessageBox(thandle,'No slew below horizon','Telescope Control',MB_SYSTEMMODAL);
     goto endit;
   End;

//
   If((TargetQuadrant = 1) And (Quadrant = 1) ) Then    GoQuad1to1(Xcount,Ycount)
   else If((TargetQuadrant = 3) And (Quadrant = 1) ) Then    GoQuad1to3(Xcount,Ycount)
   else If((TargetQuadrant = 3) And (Quadrant = 3) ) Then    GoQuad1to3(Xcount,Ycount)
   else If((TargetQuadrant = 1) And (Quadrant = 3) ) Then    GoQuad3to1(Xcount,Ycount)
   else
     begin
      Coord(pndomem^).RAGoTo := 0.0   ;
      Coord(pndomem^).decGoTo := 0.0  ;
      windows.MessageBox(thandle,'Slew below pole not yet implemented','Telescope Control',MB_SYSTEMMODAL);

     end;
endit:

//   If((TargetQuadrant = 4) And (Quadrant = 3) ) Then    GoQuad3to4 ;
//   If((TargetQuadrant = 3) And (Quadrant = 4) ) Then    GoQuad4to3 ;
//   If((TargetQuadrant = 4) And (Quadrant = 4) ) Then    GoQuad4to4 ;
//   If((TargetQuadrant = 1) And (Quadrant = 2) ) Then    GoQuad1to1 ;
//   If((TargetQuadrant = 2) And (Quadrant = 2) ) Then    GoQuad1to1 ;
//   If((TargetQuadrant = 2) And (Quadrant = 1) ) Then    GoQuad1to1 ;


End;
end.

File: GVO Code C++ v1/Slew.cpp
#include "Slew.h"

#include <cmath>
#include <iostream>
#include <windows.h>

#include "CommUtils.h"
#include "GetQandYU.h"
#include "GlobalValues.h"
#include "quad1to1.h"
#include "quad1to3.h"
#include "quad3to1.h"
#include "quad3to3.h"

void SlewScope(double raTarget, double decTarget) {
	double Alt = 0.0;
	double HA = 0.0;
	double Xcount = 0.0;
	double Ycount = 0.0;
	int localTargetQuadrant = 0;
	int localTargetYPole = 0;

	std::cout << "SlewScope: request received "
			  << "current quadrant=" << quadrant
			  << " current yPole=" << yPole
			  << " target RA=" << raTarget
			  << " target Dec=" << decTarget
			  << "\n";

	GetQandY(
		raTarget,
		decTarget,
		Alt,
		HA,
		Xcount,
		Ycount,
		localTargetQuadrant,
		localTargetYPole
	);

	std::cout << "SlewScope: target solution "
			  << "targetQuadrant=" << localTargetQuadrant
			  << " targetYPole=" << localTargetYPole
			  << " Alt=" << Alt
			  << " HA=" << HA
			  << " Xcount=" << Xcount
			  << " Ycount=" << Ycount
			  << "\n";

	if (Alt < 20.0 || localTargetYPole == 0) {
		if (Coord* coord = CommUtils::GetCoordPtr()) {
			coord->RAGoto = 0.0;
			coord->DecGoto = 0.0;
		}
		std::cout << "SlewScope: No slew, target below horizon or invalid\n";
		return;
	}

	bool slewStarted = false;

	if (localTargetQuadrant == 1 && quadrant == 1) {
		std::cout << "SlewScope: executing GoQuad1to1\n";
		slewStarted = GoQuad1to1(Xcount, Ycount);
	}
	else if (localTargetQuadrant == 3 && quadrant == 1) {
		std::cout << "SlewScope: executing GoQuad1to3\n";
		slewStarted = GoQuad1to3(Xcount, Ycount, decTarget);
	}
	else if (localTargetQuadrant == 3 && quadrant == 3) {
		std::cout << "SlewScope: executing GoQuad1to3\n";
		slewStarted = GoQuad1to3(Xcount, Ycount, decTarget);
	}
	else if (localTargetQuadrant == 1 && quadrant == 3) {
		std::cout << "SlewScope: executing GoQuad3to1\n";
		slewStarted = GoQuad3to1(Xcount, Ycount, decTarget);
	}
	else {
		if (Coord* coord = CommUtils::GetCoordPtr()) {
			coord->RAGoto = 0.0;
			coord->DecGoto = 0.0;
		}
	}

	if (slewStarted) {
		// Pascal behavior: the quad routine marks moving true and returns.
		if (Coord* coord = CommUtils::GetCoordPtr()) {
			coord->RAGoto = 0.0;
			coord->DecGoto = 0.0;
		}

		std::cout << "SlewScope: slew started\n";
	}
	else {
		std::cout << "SlewScope: slew failed to send\n";
	}
}